Document blockchain.scripthash.first_use

dagurval opened the pull request:
I think there should be room to document RPC calls, that we've yet to agree on if go into the main protocol.

This adds a new section "Methods Extra". 

To quote the documentation in this PR:
> This attempts to document implementation specific extensions that are intended
> to be widely useful. These are not part of the main specification, but may be
> suitable for eventual standardization.
> 
> If the server supports any of these extensions, these must be discoverable via
> the RPC call `server.features`.

I start by documenting an upcoming feature in ElectrsCash:: https://github.com/BitcoinUnlimited/ElectrsCash/pull/34

dagurval committed:
Document blockchain.*.first_use

cculianu commented:
Initially I didn't understand what UI/usecase in a client would be for this -- but after talking to you in Telegram @dagurval  it sounds like it would help some auto-derivation-path discovery code work a lot better.

On the surface this looks good.  I will look at it more carefully a bit later today or tomorrow (like the exact wording used if there are clarification that are needed) -- but on first read it looks good.

cculianu commented:
Hi,

Ok so I envision this working as follows for Electron Cash --

1. User inputs some BIP39 seed
2. User has no idea what the derivation path for that seed was (this is a VERY common predicament for users judging by what questions users ask us on Telegram)
3. Electron Cash then "probes" a bunch of common-used paths by submitting a batch of 20 scripthashes from each path in its hard-coded list of "paths to try".
4. If any of the path(s) have a history, user is presented with a UI option to pick one of the paths (or all of them) to create wallet(s) for them.

In that case -- the GUI doesn't really need the txhash info and all the other extra stuff.  It just cares about whether an answer to the question "does SH x have a history?" is  a boolean true/false.

I was thinking it would be faster for some server implementations (ElectrumX or Fulcrum work this way) to not have to return extra data if the client doesn't need it.  Some clients may only want the `true`/`false` answer to the question "does this scripthash X have a history?".

In that regard -- how do you feel about an additional **optional** (?) parameter to the method, "verbose".

In the verbose case it works as described.

In the non-verbose case perhaps it just returns a simple boolean `true` / `false`.

The motivation for this is that on Fulcrum it would be faster to just return a boolean "exist" / "not-exit" state rather than the full data -- especially if the client doesn't require the data.  It would be a slight polite optimization on the part of the client to be specific about what level of detail it needs. :)

Alternatively -- such a call can be supported by a different method, say `blockchain.scripthash.has_history` or something like that.

cculianu commented:
I have concerns about the motivation for this:

> Returns the hash and height of the first block scripthash occured in on the
block chain. It also includes a list of all transaction in that block that
contain the scripthash in its output.

What is the usefulness of having an incomplete list of transactions in the block that contain this scripthash as an output, while *omitting* the tx's that contain it as only an input?

In my mind you might as well just not return a list then -- if the list is potentially incomplete.  What is the purpose of this incomplete list of tx's in terms of the client-side?

I can't think of any reason why the client would want this list offhand.  Can you illuminate me?

In-between those **output** tx's, potentially, there could be many tx's involving the scripthash as input -- but they don't appear in the resulting list!  

What is the purpose of having an **incomplete** list of tx's for that block?

Wouldn't it be better to just return 1 tx involving that sh and be done with it?  (In that case you have to decide what is the canonical "first" tx -- is it the first tx in blockchain order?).

Motivation for this:

For ElectrumX and for Fulcrum -- this requirement that it be **just** a list of tx's where SH appeared in the outputs (but NOT where it was only in the tx inputs) is actually going to negatively impact performance.  

Why? I'll tell you why:

In both implementations: histories are stored in such a way where **whether** a particular tx was an input or an output for a scripthash is not stored in the db.. only the fact that the tx "happened" and "involved SH x" is stored in the history.

For Fulcrum or ElectrumX to implement this 100% to-spec -- they would need to be able to differentiate the two (and filter out tx's where the scripthash was only in the inputs).  That would involve redoing the data model of the db -- or involve slow calls to bitcoind.

I strongly suggest simplifying this to just return a block height or something or a tx -- but not an incomplete list of *output-only* tx's.

This sounds like an implementation detail that benefits Electrs but has negative consequence fo other potential implementations.

My suggestion:

1. Lift the requirement that the entire list be *just* outputs and allow for **outputs** and/or **outputs & inputs**

*or*

2. Just return the first tx seen for a SH in the block... in blockchain order.

*or*

3. Don't return tx's at all and just return a block height and/or a block hash (or both).

I presume the reason you want the tx and the block header hash is for some merkle-proofing or something?  I think returning any arbitrary tx in the block that involves SH x should be sufficient for that purpose.

cculianu commented:
Last comment:  I would appreciate some discussion or wallet use-cases or *something* that motivated the particular design of this method as it stands.

Thanks.

dagurval commented:
> In that regard -- how do you feel about an additional optional (?) parameter to the method, "verbose".

Works for me

> Alternatively -- such a call can be supported by a different method, say blockchain.scripthash.has_history or something like that.

I like that even better

(https://github.com/BitcoinUnlimited/ElectrsCash/pull/34 has discussion on the other stuff)

cculianu commented:
Yeah ok, I would go with `has_history` too then.

cculianu commented:
Ok, well since this will remain an Electrs-only feature for now -- then go ahead and merge it.

I think for Electron Cash we will offer a simplified version.. `has_history`.. returns a boolean.  May specify up to 20 or 100 address batches.  It will make discovery of paths quick.

Later on we will do a real merkle proof phase to "prove not lying".

Anyway -- you may merge this -- we definitely need a section like this anyway in the spec.

cculianu reviewed (approved):
While I disagree with the design of the 1 proposed method in this section.. I do agree with the existence of this section and i'm glad it's broken out into an "extra" section.

dagurval commented:
Thanks @cculianu, I value your feedback. 

Based on the discussion, I want to change it to only return one transaction before merging. 

I also learned about the internals of Fulcrum and ElectrumX, which is useful for thinking of future extensions. Not relevant, but I'm guessing this function would be trivial to implement on these implementations on blockchains with TTOR.

The "extra" section is indeed useful for things we agree to disagree on :-)

cculianu commented:
>  blockchains with TTOR. 

Yes, right.  Then the output tx is always first.

Ok well.. I didn't mean to cause a big stir.  I figured if we are specifying methods, it's good to weigh the contradictory opinions.  

Having the spec say: "Returns at least 1, possibly more than 1 output tx" (an arbitrary output tx in the block) would also be a relaxation of the spec.

That being said for the way Electron Cash works -- it wouldn't even use the tx info for a real "not lying" style proof.  

Electron Cash would rather have that boolean data of "does it have a history??".

Later on when the user selects from the path(s) to create wallet(s) -- it would then do a real download of the full history, along with merkle proof magic.

So for Electron Cash I think I'll add the `has_history` style call later on when we get to the point where we implement auto-path-discovery.

dagurval commented:
The following changes were added:
* Spec simplified to returning just on tx hash. The tx hash can be any transaction containing scripthash as output.
* Added rationale for requiring scripthash to be output
* Added `blockchain.address.get_first_use`
